File: src/stores/product.ts
import { create } from 'zustand';
import { persist } from 'zustand/middleware';

import { protectedApi } from '../lib/al/axios';
import type { Product } from '../types/product';

export interface ProductData extends Product {}

interface ProductState {
   products: ProductData[];

   all: (params?: any) => Promise<any>;
   getCatalog: (params?: any) => Promise<any>;
   detail: ({ id }: { id: string }) => Promise<any>;
   add: ({ data }: { data: ProductData }) => Promise<any>;
   update: ({ id, data }: { id: string; data: ProductData }) => Promise<any>;
   delete: ({ id }: { id: string }) => Promise<any>;
}

const useProductStore = create<ProductState>()(
   persist(
      (set, get) => ({
         products: [],
         all: async (params?: any) => {
            const queryParams = new URLSearchParams();

            if (params) {
               if (params.page !== undefined) queryParams.append('page', params.page.toString());
               if (params.limit !== undefined) queryParams.append('limit', params.limit.toString());
               if (params.search !== undefined) queryParams.append('search', params.search);
               if (params.sort !== undefined) queryParams.append('sort', params.sort);
               if (params.order !== undefined) queryParams.append('order', params.order);
               if (params.product_type_id !== undefined)
                  queryParams.append('product_type_id', params.product_type_id);
               if (params.is_active !== undefined)
                  queryParams.append('is_active', params.is_active);
            }

            const queryString = queryParams.toString();
            const url = queryString ? `/products?${queryString}` : '/products';

            const response = await protectedApi.get(url);
            if (response.data.success && response.data.data) {
               set({ products: response.data.data.products || response.data.data });
            }
            return response.data;
         },
         getCatalog: async (params?: any) => {
            const queryParams = new URLSearchParams();

            if (params) {
               if (params.page !== undefined) queryParams.append('page', params.page.toString());
               if (params.limit !== undefined) queryParams.append('limit', params.limit.toString());
               if (params.search !== undefined) queryParams.append('search', params.search);
               if (params.sort !== undefined) queryParams.append('sort', params.sort);
               if (params.order !== undefined) queryParams.append('order', params.order);
               if (params.product_type_id !== undefined)
                  queryParams.append('product_type_id', params.product_type_id);
               if (params.product_type_ids !== undefined)
                  queryParams.append('product_type_ids', params.product_type_ids);
            }

            const queryString = queryParams.toString();
            const url = queryString ? `/catalog?${queryString}` : '/catalog';

            const response = await protectedApi.get(url);
            if (response.data.success && response.data.data) {
               set({ products: response.data.data.products || response.data.data });
            }
            return response.data;
         },
         detail: async ({ id }) => {
            const response = await protectedApi.get(`/products/${id}`);
            return response.data;
         },
         add: async ({ data }) => {
            let payload: any = data;
            let config = {};

            // If image is a File, use FormData
            if (data.image instanceof File) {
               const formData = new FormData();

               if (data.product_number) formData.append('product_number', data.product_number);
               if (data.product_type_id) formData.append('product_type_id', data.product_type_id);
               if (data.title) formData.append('title', data.title);
               if (data.slug) formData.append('slug', data.slug);
               if (data.description) formData.append('description', data.description);
               if (data.sale_price !== undefined)
                  formData.append('sale_price', data.sale_price.toString());
               if (data.content) formData.append('content', data.content);
               if (data.is_active !== undefined)
                  formData.append('is_active', data.is_active.toString());
               if (data.tracking_mode) formData.append('tracking_mode', data.tracking_mode);
               if (data.measurement_unit)
                  formData.append('measurement_unit', data.measurement_unit);
               if (data.image instanceof File) {
                  formData.append('image', data.image);
               }

               payload = formData;
               config = {
                  headers: {
                     'Content-Type': 'multipart/form-data',
                  },
               };
            }

            const response = await protectedApi.post(`/products`, payload, config);
            return response.data;
         },
         update: async ({ id, data }) => {
            let payload: any = data;
            let config = {};

            // If image is a File, use FormData
            if (data.image instanceof File) {
               const formData = new FormData();

               if (data.product_number) formData.append('product_number', data.product_number);
               if (data.product_type_id) formData.append('product_type_id', data.product_type_id);
               if (data.title) formData.append('title', data.title);
               if (data.slug) formData.append('slug', data.slug);
               if (data.description) formData.append('description', data.description);
               if (data.sale_price !== undefined)
                  formData.append('sale_price', data.sale_price.toString());
               if (data.content) formData.append('content', data.content);
               if (data.is_active !== undefined)
                  formData.append('is_active', data.is_active.toString());
               if (data.tracking_mode) formData.append('tracking_mode', data.tracking_mode);
               if (data.measurement_unit)
                  formData.append('measurement_unit', data.measurement_unit);
               if (data.image instanceof File) {
                  formData.append('image', data.image);
               }

               payload = formData;
               config = {
                  headers: {
                     'Content-Type': 'multipart/form-data',
                  },
               };
            }

            const response = await protectedApi.post(`/products/${id}`, payload, config);
            return response.data;
         },
         delete: async ({ id }) => {
            const response = await protectedApi.delete(`/products/${id}`);
            return response.data;
         },
      }),
      {
         name: 'product-store',
         partialize: (state) => ({
            products: state.products,
         }),
      }
   )
);

export default useProductStore;

File: modules/cms/handlers/product.go
package handlers

import (
	"aldev/modules/cms/models"
	"aldev/utils"
	"fmt"
	"strconv"
	"strings"

	"github.com/go-playground/validator/v10"
	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type ProductInput struct {
	ProductNumber   *string    `json:"product_number" validate:"required,max=100"`
	ProductTypeID   *uuid.UUID `json:"product_type_id" validate:"required"`
	Title           *string    `json:"title" validate:"required,max=300"`
	Slug            *string    `json:"slug" validate:"required,max=300"`
	Image           *string    `json:"image,omitempty"`
	Description     *string    `json:"description,omitempty"`
	SalePrice       *float64   `json:"sale_price" validate:"required,gt=0"`
	Content         *string    `json:"content,omitempty"`
	IsActive        *bool      `json:"is_active"`
	TrackingMode    *string    `json:"tracking_mode,omitempty"`
	MeasurementUnit *string    `json:"measurement_unit,omitempty"`
}

type ProductHandler struct {
	DB *gorm.DB
}

func NewProductHandler(db *gorm.DB) *ProductHandler {
	return &ProductHandler{DB: db}
}

func (h *ProductHandler) GetProduct(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return utils.RespApi(c, "bad", "Id yang diberikan tidak valid", nil)
	}

	var product models.Product
	if err := h.DB.Preload("ProductType").First(&product, "id = ?", id).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal mendapatkan data Product", err.Error())
	}

	return utils.RespApi(c, "ok", "Berhasil mendapatkan data Product", product)
}

func (h *ProductHandler) GetAllProducts(c *fiber.Ctx) error {
	page, _ := strconv.Atoi(c.Query("page", "1"))
	limit, _ := strconv.Atoi(c.Query("limit", "10"))
	search := strings.ToLower(c.Query("search", ""))
	sort := c.Query("sort", "id")
	order := c.Query("order", "asc")
	productTypeID := c.Query("product_type_id", "")
	isActive := c.Query("is_active", "")

	offset := (page - 1) * limit

	db := h.DB.Model(&models.Product{}).Preload("ProductType")

	// Filter search
	if search != "" {
		db = db.Where(`
			LOWER(title) LIKE ? OR 
			LOWER(description) LIKE ? OR
			LOWER(product_number) LIKE ?`,
			"%"+search+"%", "%"+search+"%", "%"+search+"%",
		)
	}

	// Filter by product type
	if productTypeID != "" {
		db = db.Where("product_type_id = ?", productTypeID)
	}

	// Filter by active status
	if isActive != "" {
		if isActive == "true" {
			db = db.Where("is_active = ?", true)
		} else if isActive == "false" {
			db = db.Where("is_active = ?", false)
		}
	}

	// Count total
	var total int64
	if err := db.Count(&total).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal hitung total", err.Error())
	}

	// Sorting (whitelisted)
	validSortFields := map[string]string{
		"id":             "id",
		"title":          "title",
		"product_number": "product_number",
		"sale_price":     "sale_price",
		"stock":          "stock",
		"created_at":     "created_at",
	}
	sortBy, ok := validSortFields[sort]
	if !ok {
		sortBy = "id"
	}
	db = db.Order(fmt.Sprintf("%s %s", sortBy, order))

	// Fetch data
	var products []models.Product
	if err := db.Offset(offset).Limit(limit).Find(&products).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal ambil data", err.Error())
	}

	totalPages := (total + int64(limit) - 1) / int64(limit)

	result := fiber.Map{
		"products": products,
		"pagination": fiber.Map{
			"total":      total,
			"page":       page,
			"limit":      limit,
			"totalPages": totalPages,
		},
	}

	return utils.RespApi(c, "ok", "Berhasil mendapatkan data Products", result)
}

func (h *ProductHandler) AddProduct(c *fiber.Ctx) error {
	var input ProductInput

	contentType := c.Get("Content-Type")
	if strings.Contains(contentType, "multipart/form-data") {
		// Handle multipart form data for file uploads
		productNumber := c.FormValue("product_number")
		productTypeIDStr := c.FormValue("product_type_id")
		title := c.FormValue("title")
		slug := c.FormValue("slug")
		description := c.FormValue("description")
		salePriceStr := c.FormValue("sale_price")
		content := c.FormValue("content")
		isActiveStr := c.FormValue("is_active")

		// Parse UUID
		productTypeID, err := uuid.Parse(productTypeIDStr)
		if err != nil {
			return utils.RespApi(c, "bad", "Product Type ID tidak valid", err.Error())
		}

		// Parse float
		salePrice, err := strconv.ParseFloat(salePriceStr, 64)
		if err != nil {
			return utils.RespApi(c, "bad", "Sale Price tidak valid", err.Error())
		}

		// Parse bool
		isActive := true
		if isActiveStr == "false" {
			isActive = false
		}

		// Parse tracking fields
		trackingMode := c.FormValue("tracking_mode")
		measurementUnit := c.FormValue("measurement_unit")

		input = ProductInput{
			ProductNumber: &productNumber,
			ProductTypeID: &productTypeID,
			Title:         &title,
			Slug:          &slug,
			Description:   &description,
			SalePrice:     &salePrice,
			Content:       &content,
			IsActive:      &isActive,
		}

		if trackingMode != "" {
			input.TrackingMode = &trackingMode
		}
		if measurementUnit != "" {
			input.MeasurementUnit = &measurementUnit
		}

		// Handle file upload
		if file, err := c.FormFile("image"); err == nil && file != nil {
			filePath, err := utils.UploadFile(c, "image", "products")
			if err != nil {
				return utils.RespApi(c, "bad", "Gagal upload image Product", err.Error())
			}
			input.Image = &filePath
		}
	} else {
		// Handle JSON data
		if err := c.BodyParser(&input); err != nil {
			return utils.RespApi(c, "bad", "Request Body tidak valid", err.Error())
		}
	}

	if err := utils.Validate.Struct(input); err != nil {
		if verrs, ok := err.(validator.ValidationErrors); ok {
			return utils.RespApi(c, "bad", "Validasi gagal", verrs.Translate(utils.Translator))
		}
		return utils.RespApi(c, "bad", "Validasi gagal", err.Error())
	}

	// Initialize stock to 0
	initialStock := 0

	product := models.Product{
		ProductNumber:   input.ProductNumber,
		ProductTypeID:   input.ProductTypeID,
		Title:           input.Title,
		Slug:            input.Slug,
		Description:     input.Description,
		Image:           input.Image,
		SalePrice:       input.SalePrice,
		Content:         input.Content,
		IsActive:        input.IsActive,
		Stock:           &initialStock,
		TrackingMode:    input.TrackingMode,
		MeasurementUnit: input.MeasurementUnit,
	}

	if err := h.DB.Create(&product).Error; err != nil {
		return utils.RespApi(c, "ise", "Tidak dapat membuat Product", err.Error())
	}

	// Load ProductType relation
	h.DB.Preload("ProductType").First(&product, "id = ?", product.ID)

	return utils.RespApi(c, "ok", "Berhasil membuat data Product", product)
}

func (h *ProductHandler) UpdateProduct(c *fiber.Ctx) error {
	idParam := c.Params("id")
	id, err := uuid.Parse(idParam)
	if err != nil {
		return utils.RespApi(c, "bad", "ID yang diberikan tidak valid", nil)
	}

	var product models.Product
	if err := h.DB.First(&product, "id = ?", id).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal Mendapatkan Product", err.Error())
	}

	var input ProductInput

	contentType := c.Get("Content-Type")
	if strings.Contains(contentType, "multipart/form-data") {
		// Handle multipart form data for file uploads
		productNumber := c.FormValue("product_number")
		productTypeIDStr := c.FormValue("product_type_id")
		title := c.FormValue("title")
		slug := c.FormValue("slug")
		description := c.FormValue("description")
		salePriceStr := c.FormValue("sale_price")
		content := c.FormValue("content")
		isActiveStr := c.FormValue("is_active")

		// Parse UUID
		productTypeID, err := uuid.Parse(productTypeIDStr)
		if err != nil {
			return utils.RespApi(c, "bad", "Product Type ID tidak valid", err.Error())
		}

		// Parse float
		salePrice, err := strconv.ParseFloat(salePriceStr, 64)
		if err != nil {
			return utils.RespApi(c, "bad", "Sale Price tidak valid", err.Error())
		}

		// Parse bool
		isActive := true
		if isActiveStr == "false" {
			isActive = false
		}

		// Parse tracking fields
		trackingMode := c.FormValue("tracking_mode")
		measurementUnit := c.FormValue("measurement_unit")

		input = ProductInput{
			ProductNumber: &productNumber,
			ProductTypeID: &productTypeID,
			Title:         &title,
			Slug:          &slug,
			Description:   &description,
			SalePrice:     &salePrice,
			Content:       &content,
			IsActive:      &isActive,
		}

		if trackingMode != "" {
			input.TrackingMode = &trackingMode
		}
		if measurementUnit != "" {
			input.MeasurementUnit = &measurementUnit
		}

		// Handle file upload
		if file, err := c.FormFile("image"); err == nil && file != nil {
			oldImagePath := ""
			if product.Image != nil {
				oldImagePath = *product.Image
			}

			filePath, err := utils.UpdateFile(c, oldImagePath, "image", "products")
			if err != nil {
				return utils.RespApi(c, "bad", "Gagal memperbarui image Product", err.Error())
			}
			input.Image = &filePath
		} else {
			// If no new file uploaded, keep the old image
			input.Image = product.Image
		}
	} else {
		// Handle JSON data
		if err := c.BodyParser(&input); err != nil {
			return utils.RespApi(c, "bad", "Request Body tidak valid", err.Error())
		}

		// Sanitize image path - remove URL prefix if present
		if input.Image != nil && *input.Image != "" {
			imagePath := *input.Image
			// Remove http:// or https:// prefix and domain
			if strings.HasPrefix(imagePath, "http://") || strings.HasPrefix(imagePath, "https://") {
				// Extract path starting from "uploads/"
				if idx := strings.Index(imagePath, "uploads/"); idx != -1 {
					sanitizedPath := imagePath[idx:]
					input.Image = &sanitizedPath
				}
			}
		}
	}

	if err := utils.Validate.Struct(input); err != nil {
		if verrs, ok := err.(validator.ValidationErrors); ok {
			return utils.RespApi(c, "bad", "Validasi gagal", verrs.Translate(utils.Translator))
		}
		return utils.RespApi(c, "bad", "Validasi gagal", err.Error())
	}

	updates := map[string]interface{}{
		"product_number":   input.ProductNumber,
		"product_type_id":  input.ProductTypeID,
		"title":            input.Title,
		"slug":             input.Slug,
		"description":      input.Description,
		"image":            input.Image,
		"sale_price":       input.SalePrice,
		"content":          input.Content,
		"is_active":        input.IsActive,
		"tracking_mode":    input.TrackingMode,
		"measurement_unit": input.MeasurementUnit,
	}

	if err := h.DB.Model(&product).Updates(updates).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal memperbarui data Product", err.Error())
	}

	// Fetch updated data with ProductType
	if err := h.DB.Preload("ProductType").First(&product, "id = ?", id).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal mendapatkan data terbaru", err.Error())
	}

	return utils.RespApi(c, "ok", "Berhasil memperbarui data Product", product)
}

func (h *ProductHandler) DeleteProduct(c *fiber.Ctx) error {
	idParam := c.Params("id")
	id, err := uuid.Parse(idParam)
	if err != nil {
		return utils.RespApi(c, "bad", "ID yang diberikan tidak valid", nil)
	}

	var product models.Product
	if err := h.DB.First(&product, "id = ?", id).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal Mendapatkan Product", err.Error())
	}

	// Soft delete
	if err := h.DB.Delete(&product).Error; err != nil {
		return utils.RespApi(c, "ise", "Gagal Menghapus Product", err.Error())
	}

	return utils.RespApi(c, "ok", "Berhasil Menghapus Product", nil)
}
